src/CPU/ctxswitchkernel.cpp

```diff
@@ -0,0 +1,118 @@
+#include <iostream>
+#include <chrono>
+#include <thread>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <sched.h>
+#include <pthread.h>
+#include <stdexcept>
+#include "timer.hpp"
+#include "stats.hpp"
+#include "proc.hpp"
+
+const int pipes = 20; // number of pipes or processes in ring
+const int footprint_vector_len = 100000; // cache footprint
+const int iters = 100;
+
+using namespace std;
+
+void *overhead_routine(void *ptr) {
+    // create pipes
+    use_cores(vector<int> {1});
+    int (*p)[2] = (int (*)[2]) ptr;
+    char buf = 'm';
+    for (size_t i = iters; i > 0; i--) {
+        for (int i = 0; i < pipes; i++) {
+            read(p[i][0], &buf, 1); // read from pipe
+            write(p[(i + 1) % pipes][1], &buf, 1); // write to pipe
+        }
+    }
+    return NULL;
+}
+
+
+
+double overhead() {
+    
+    int p[pipes][2];
+    pthread_t tid;
+    char buf = 'm';
+
+    for (int i = 0; i < pipes; i++) {
+        if (pipe(p[i]) == -1) {
+            throw std::runtime_error("cant create pipe");
+        }
+    }
+    pthread_create(&tid, NULL, overhead_routine, (void *) p);
+
+    Timer t;
+    t.begin();
+    write(p[0][1], &buf, 1); // write to pipe
+    pthread_join(tid, NULL);
+    t.end();
+
+    for (int i = 0; i < pipes; i++) {
+        close(p[i][0]);
+        close(p[i][1]);
```

Comment:
Not affecting the experiment... but can we do this for ctxswitch.cpp also... I guess we are missing closing the write end of the pipe there.

Reply:
yup, I'll add this.